tests/client/wire_protocol.rs

```diff
@@ -0,0 +1,40 @@
+use bson::Document;
+use bson::Bson::FloatingPoint;
+use mongodb::client::wire_protocol::operations::Message;
+use std::io::Write;
+use std::net::TcpStream;
+
+// Not prefixing this with "#[test]" since it requires both a server to be
+// running locally and the correct database state to pass, and we don't want
+// Travis to fail.
+fn wire_protocol() {
+    match TcpStream::connect("localhost:27017") {
+        Ok(mut stream) => {
+            println!("Successfully connected to server");
+
+            let doc = Document::new();
+            let res = Message::new_query(1, 0, "test.test".to_owned(), 0, 1, doc, None);
+
+            let cm = match res {
+                Ok(message) => message,
+                Err(_) => panic!("Could not create message!")
+            };
+
+            match cm.write(&mut stream) {
+                Ok(_) => (),
+                Err(s) => panic!("{}", s)
+            };
+
+            match Message::read(&mut stream) {
+                Ok(bson) => match bson.get("foo") {
```

Comment:
nested matches are not great for readability + maintainability, I'd put this behind a separate variable